BrowserServicesKit/Sources/Suggestions/SuggestionProcessing.swift

```diff
@@ -42,263 +50,241 @@ final class SuggestionProcessing {
+        let duckDuckGoPhrases = (apiResult?.items ?? [])
+            .compactMap { suggestion -> Suggestion? in
+                guard suggestion.isNav != true, let phrase = suggestion.phrase else { return nil }
+                return Suggestion.phrase(phrase: phrase)
+            }
+            .prefix(Self.maximumNumberOfSuggestions - (topHits.count + historyAndBookmarksAndOpenTabs.count))
+            .map(\.self)
+
+        // STEP 11: Return final ordered suggestions
+        return SuggestionResult(
+            topHits: topHits,
+            duckduckgoSuggestions: duckDuckGoPhrases,
+            localSuggestions: historyAndBookmarksAndOpenTabs
+        )
     }
 
-    private func dedupLocalSuggestions(_ suggestions: [Suggestion]) -> [Suggestion] {
-        return suggestions.reduce([]) { partialResult, suggestion in
-            if partialResult.contains(where: {
-
-                switch $0 {
-                case .bookmark(title: let title, url: let url, isFavorite: let isFavorite, allowedInTopHits: _):
-                    if case .bookmark(let searchTitle, let searchUrl, let searchIsFavorite, _) = suggestion,
-                       searchTitle == title,
-                       searchUrl.naked == url.naked,
-                       searchIsFavorite == isFavorite {
-                        return true
-                    }
-
-                case .historyEntry(title: let title, url: let url, allowedInTopHits: _):
-                    if case .historyEntry(let searchTitle, let searchUrl, _) = suggestion,
-                       searchTitle == title,
-                       searchUrl.naked == url {
-                        return true
-                    }
-
-                case .internalPage(title: let title, url: let url):
-                    if case .internalPage(let searchTitle, let searchUrl) = suggestion,
-                       searchTitle == title,
-                       searchUrl == url {
-                        return true
-                    }
-
-                case .openTab(title: let title, url: let url, tabId: _):
-                    if case .openTab(let searchTitle, let searchUrl, _) = suggestion,
-                       searchTitle == title,
-                       searchUrl.naked == url.naked {
-                        return true
-                    }
-
-                default:
-                    assertionFailure("Unexpected suggestion in local suggestions")
-                    return true
-                }
-
-                return false
-            }) {
-                return partialResult
+    /// Removes duplicate entries (based on the URL) from a list of suggestions.
+    /// When duplicates are found, ones with more info (e.g. bookmarks) will take precedence.
+    private func removeDuplicates(_ suggestions: [ScoredSuggestion]) -> [(suggestion: ScoredSuggestion, kinds: Set<ScoredSuggestion.Kind>)] {
+        // Group suggestions by normalized URL preserving the keys order
```

Comment:
the deduplication algorithm changed to match the windows implementation with optimized performance (dict grouping instead of O² of the nested array) + key order preservation for stable testability. Also we “duplicate” a non-open-tab suggestion in case it gets to the Top Hits section so we present both the history and Switch to Tab option in the suggestions – see the comments inline